### DISCUSSION TITLE
Vue Test Utils Mock

### DISCUSSION THREAD

**Discussion by T0miii** [urql-graphql/urql#2125] [Q&A]
Hello,

is there an example how to mock the Urql client for jest with the Vue Implementation? 

ty

**Comment by kitten** [comment]
Hiya :wave:

All of these suggestions from the Testing guide apply to any framework, since the `Client` is framework-agnostic and part of `@urql/core`: https://formidable.com/open-source/urql/docs/advanced/testing/
The only thing you'd have to change is that `Provider` becomes `provideClient` or a similar installation for Vue.

**Comment by T0miii** [comment]
Ty for some reason provideClient didnt work for me, so i did it like this in case some one else faces the same issue:
```
            global: {
                plugins: [[urql, {
                    url: "fakeUrl",
                }]],
```

**Comment by bernardo-zuchowski** [comment]
Hi guys, I've been trying to mock the client and I haven't had any success. I couldn't understand how to replace `Provider` with `provideClient`. Any example for Vue 3 besides the one already in the documentation?

**Comment by lucastraba** [comment]
Hey guys! I'd like to also bump this, because I've spent the last couple of hours and I cannot make it work.

If you do exactly as shown in the docs and pointed out above, it looks something like this:

```ts
const mockClient = {
    executeQuery: vi.fn(() => never),
    executeMutation: vi.fn(() => never),
    executeSubscription: vi.fn(() => never),
  };

provideClient(mockClient);
```

This causes this error:
`Error: You are creating an urql-client without a url.`

Adding the url, thus:
```ts
const mockClient = {
    url: 'foo',
    executeQuery: vi.fn(() => never),
    executeMutation: vi.fn(() => never),
    executeSubscription: vi.fn(() => never),
  };

provideClient(mockClient);
```

Causes the error:
`TypeError: Cannot read properties of undefined (reading 'reduceRight')`

Is there anything we're missing?

Also, @T0miii's solution didn't work for me, provided that where it says `urql` it is the mock object that needs to feed the provider. What is `urql` in your example?

Thanks a bunch

**Comment by dani-sc** [comment]
Hi guys,
did any of you (@bernardo-zuchowski , @lucastraba, @kitten  ) ever figure out how to mock the urql client in Vue?

**Reply by lucastraba** [reply]
Nope, unfortunately... Hope the guys get back to us on this one soon!

**Reply by bernardo-zuchowski** [reply]
Sorry, I didn't know how to figure it out.

**Reply by SeJunB** [reply]
I was able to mock the urql client by:
1. Create the mock client
```
import { never } from "wonka";
// Must be a ref.
export const mockUrqlClient = ref({
  executeQuery: vi.fn(() => never),
  executeMutation: vi.fn(() => never),
});
```
2. Provide the mock client by `config.global.provide = { "$urql": mockUrqlClient};`. The key must be "$urql" since [useClient](https://github.com/urql-graphql/urql/blob/main/packages/vue-urql/src/useClient.ts#L100), the method that `useQuery`/`useMutation` use to get the client, uses `inject('$urql')` to get the client.

Please see https://github.com/SeJunB/urql-client-unit-test for an example for a complete example.

FYI, if you want your mock to return a response, ensure that the object that you pass into the `fromValue` function is in the following format:
```
{
  data: {/* your data */}
}
```

**Reply by lucastraba** [reply]
You, sir or madam, are a savior of the broken, the beaten and the damned 🔥
